src/Interpreters/Cache/FileCache.cpp

```diff
@@ -456,32 +494,48 @@ FileSegmentsHolder FileCache::get(const Key & key, size_t offset, size_t size)
 }
 
 FileCache::FileSegmentCell * FileCache::addCell(
-    const Key & key, size_t offset, size_t size,
-    FileSegment::State state, const CreateFileSegmentSettings & settings,
+    const Key & key,
+    size_t offset,
+    size_t size,
+    FileSegment::State state,
+    const CreateFileSegmentSettings & settings,
     std::lock_guard<std::mutex> & cache_lock)
 {
     /// Create a file segment cell and put it in `files` map by [key][offset].
 
     if (!size)
         return nullptr; /// Empty files are not cached.
 
-    if (files[key].contains(offset))
+    auto & offsets = files[key].file_segments_by_offset;
+
+    if (offsets.empty())
+    {
+        auto key_path = getPathInLocalCache(key);
+
+        if (!fs::exists(key_path))
+            fs::create_directories(key_path);
```

Comment:
Make a separate method `initializeNewKey()` and add `on_key_eviction_func` there?